### PR TITLE
homework_2

### PR DESCRIPTION
**Pull request by Puvel**

### CODE REVIEW

File: src/App.jsx [hunk 1]
```diff
@@ -1,15 +1,129 @@
...
-		<Courses />
-		<ToastContainer theme='dark' />
-	</>
-);
+export const App = () => {
+	const [user, setUser] = useState(null);
+	const [isLoading, setIsLoading] = useState(false);
+	const [isLoggedIn, setIsLoggedIn] = useState(
+		!!localStorage.getItem('isLoggedIn')
+	);
```
Comment (KitTheProgrammer) [inline src/App.jsx:32]: It is better to set the initial value from `useEffect`

You have one already at line 45, you can do it there

File: src/components/CourseInfo/CourseInfo.jsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import { authorGenerator } from 'helpers/authorGenerator';
+import { COURSES_PATH } from 'constants';
+import { Context } from 'context/context';
+
+import style from './courseInfo.module.css';
+
+export const CourseInfo = () => {
+	let { courseId } = useParams();
+	const { authors, courses } = useContext(Context);
+	const course = courses.find((course) => course.id === courseId);
```
Comment (KitTheProgrammer) [inline src/components/CourseInfo/CourseInfo.jsx]: Use `useMemo` here

File: src/components/Login/Login.jsx [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+	const inputsStates = {
+		[INPUT_EMAIL_ID]: setEmail,
+		[INPUT_PASSWORD_ID]: setPassword,
+	};
+
+	const navigate = useNavigate();
+
+	const handleChange = (ukey, { target: { value } }) =>
+		inputsStates[ukey](value);
```
Comment (KitTheProgrammer) [inline src/components/Login/Login.jsx]: I understand the idea, but it feels too redundant here. You have only 2 fields, that can be set like this:
`<input onChange={({ target: { value } }) => setPassword(value)}`

File: src/components/Registration/Registration.jsx [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+	const inputsStates = {
+		[INPUT_NAME_ID]: setName,
+		[INPUT_EMAIL_ID]: setEmail,
+		[INPUT_PASSWORD_ID]: setPassword,
+	};
+
+	const handleChange = (ukey, { target: { value } }) =>
+		inputsStates[ukey](value);
+
```
Comment (KitTheProgrammer) [inline src/components/Registration/Registration.jsx]: same here

File: src/components/Registration/Registration.jsx [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+					labelId='password'
+					type={!showPass ? 'password' : 'text'}
+					labelText='Password'
+					placeholder='Enter password'
+					value={password}
+					onChange={(e) => handleChange(INPUT_PASSWORD_ID, e)}
+				/>
+
+				<button
+					onClick={() => setShowPass((prevState) => !prevState)}
```
Comment (KitTheProgrammer) [inline src/components/Registration/Registration.jsx]: usage of prev state is unnecessary

**Review by KitTheProgrammer — CHANGES_REQUESTED** [review]
1. I think this error should be present only on the registration page
image
2. Writing some random number as course id causes the app to crash (like this: /courses/12345)
image
3. You can create authors with name containing only spaces 
image
4. If you create a new course and then try to go to it's route app is crashing (by writing course/<new course id>. After reload newly created course is disappearing

**Review by EpamFanAnton — APPROVED** [review]
25

**Pull request merged**